File: libraries/di.transformer/src/diagnostics.ts
// Diagnostic plumbing for the transformer.
//
// The transformer surfaces compile-time information back to the user via the
// host's `addDiagnostic` hook (ts-patch wires this to tsc's diagnostic stream).
// We keep a thin `Diagnostic` alias over `ts.Diagnostic` plus a small set of
// stable codes so tests can assert on category + code without matching message
// text.

import ts from "typescript";

/** A diagnostic the transformer raises. Alias kept for call-site clarity. */
export type Diagnostic = ts.Diagnostic;

/** The sink the transformer writes diagnostics to (ts-patch supplies this). */
export interface DiagnosticSink {
  addDiagnostic(diagnostic: Diagnostic): number;
}

/**
 * Stable numeric codes for transformer-emitted diagnostics. The high offset
 * keeps them clear of TypeScript's own code space. These are part of the
 * transformer's observable surface — tests assert on them.
 */
export enum DiagnosticCode {
  /** A factory param's call signature doesn't match the target ctor's holes. */
  FactorySignatureMismatch = 990003,
  /**
   * A constructor / factory parameter whose type has no derivable token and
   * carries no `Inject<T, "tok">` brand — a hard compile error.
   */
  UnderivableToken = 990006,
  /**
   * A type reaches token derivation while still referencing an UNBOUND type
   * parameter — a bare generic class registered without an instantiation
   * expression (`add<IFoo<$<1>>>(Foo)` instead of `Foo<$<1>>` / `Foo<Concrete>`),
   * or a type parameter leaking into a token position. Hard compile error.
   */
  UnboundTypeParameter = 990007,
  /**
   * An open SERVICE token mixes concrete args and holes (`IFoo<$<1>,string>`,
   * `IFoo<IBar<$<1>>>`). v1 requires every type arg of an open service token to
   * be a bare hole (`IFoo<$<1>,$<2>>`; repeats like `IFoo<$<1>,$<1>>` are allowed).
   */
  MixedServiceTokenArgs = 990008,
  /**
   * An open template token on an `addValue` / factory registration. Open
   * registrations are class registrations only — a value or factory has no
   * per-closing construction the container could substitute into.
   */
  OpenTokenOnValueOrFactory = 990009,
  /**
   * A dependency slot references a hole (`$N`) that the service template does
   * not bind — substitution at close time would have no argument for it.
   */
  DepHoleNotInServiceTemplate = 990010,
  /**
   * A registration-time override array element (`add<I>(C, [...])`) is neither a
   * string-literal token nor an `undefined`/elision gap — an object literal, a
   * variable, or a call the transformer cannot resolve statically. The element is
   * ignored and the derived token is kept; use a string-literal token override.
   */
  UnresolvableOverrideElement = 990011,
}

const SOURCE = "@rhombus-std/di.transformer";

/** Build a diagnostic of `category` anchored at `node` in `file`. */
function diagnostic(
  category: ts.DiagnosticCategory,
  file: ts.SourceFile,
  node: ts.Node,
  code: DiagnosticCode,
  messageText: string,
): Diagnostic {
  return {
    file,
    start: node.getStart(file),
    length: node.getWidth(file),
    category,
    code,
    messageText,
    source: SOURCE,
  };
}

/** Build a warning diagnostic anchored at `node` in `file`. */
export function warning(
  file: ts.SourceFile,
  node: ts.Node,
  code: DiagnosticCode,
  messageText: string,
): Diagnostic {
  return diagnostic(ts.DiagnosticCategory.Warning, file, node, code, messageText);
}

/** Build an error diagnostic anchored at `node` in `file`. */
export function error(
  file: ts.SourceFile,
  node: ts.Node,
  code: DiagnosticCode,
  messageText: string,
): Diagnostic {
  return diagnostic(ts.DiagnosticCategory.Error, file, node, code, messageText);
}

File: transforms/internal/ditransform/diagnostics.go
// Package ditransform is the Go port of the registration transformer: it lowers
// the type-driven authoring forms (`addClass<I>(C)`, `.as<"x">()`, tokenless
// `resolve<I>()` / `isService<I>()`, and `nameof<T>()`) to their string-token
// runtime forms over the ttsc-shipped typescript-go checker, carrying the
// derived dependency signature inline on each registration. It is the emit-path
// twin of the hand-written TypeScript transformer; both derive identical tokens
// from the shared token core.
package ditransform

import (
	shimast "github.com/microsoft/typescript-go/shim/ast"
	shimchecker "github.com/microsoft/typescript-go/shim/checker"
	shimprinter "github.com/microsoft/typescript-go/shim/printer"

	"github.com/fnioc/std/transforms/internal/tokens"
)

// Stable diagnostic codes — part of the observable surface, asserted on by name
// rather than message text. Kept byte-identical to the reference transformer.
const (
	// codeFactorySignatureMismatch: a factory param's call signature does not
	// cover the produced constructor's caller-supplied holes.
	codeFactorySignatureMismatch = "990003"
	// codeUnderivableToken: a parameter type has no derivable token and no
	// `Inject<T,"tok">` brand.
	codeUnderivableToken = "990006"
	// codeUnboundTypeParameter: a type reaches derivation still referencing an
	// unbound type parameter (a bare generic class registered without an
	// instantiation expression).
	codeUnboundTypeParameter = "990007"
	// codeMixedServiceTokenArgs: an open service token mixes concrete args and
	// holes.
	codeMixedServiceTokenArgs = "990008"
	// codeOpenTokenOnValueOrFactory: an open template token on an addValue /
	// factory registration.
	codeOpenTokenOnValueOrFactory = "990009"
	// codeDepHoleNotInServiceTemplate: a dependency slot references a hole the
	// service template does not bind.
	codeDepHoleNotInServiceTemplate = "990010"
	// codeUnresolvableOverrideElement: a registration-time override element is
	// neither a string-literal token nor an undefined/elision gap.
	codeUnresolvableOverrideElement = "990011"
)

// Category distinguishes a hard error (fails emit) from an advisory warning.
type Category int

const (
	// Error is a hard diagnostic that fails the build.
	Error Category = iota
	// Warning is advisory and does not fail the build.
	Warning
)

// Diagnostic is one transformer-raised diagnostic destined for the sidecar
// envelope. File is the absolute declaring path; Start is the anchor node's
// position (informational — the envelope carries code + message, not position).
type Diagnostic struct {
	File     string
	Start    int
	Code     string
	Category Category
	Message  string
}

// context is the per-file lowering context: the program-wide token derivation
// context plus the per-file checker, node factory, source file, and diagnostic
// sink. It is the Go analog of the reference LowerContext.
type context struct {
	tokens  *tokens.Context
	checker *shimchecker.Checker
	factory *shimast.NodeFactory
	sf      *shimast.SourceFile
	addDiag func(Diagnostic)
	ec      *shimprinter.EmitContext
}

// emitError raises a hard diagnostic anchored at node.
func (c *context) emitError(node *shimast.Node, code, message string) {
	c.emit(Error, node, code, message)
}

// emitWarning raises an advisory diagnostic anchored at node.
func (c *context) emitWarning(node *shimast.Node, code, message string) {
	c.emit(Warning, node, code, message)
}

func (c *context) emit(category Category, node *shimast.Node, code, message string) {
	start := 0
	file := ""
	if c.sf != nil {
		file = c.sf.FileName()
	}
	if node != nil {
		start = node.Pos()
	}
	c.addDiag(Diagnostic{
		File:     file,
		Start:    start,
		Code:     code,
		Category: category,
		Message:  message,
	})
}
